In [ ]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from sklearn.preprocessing import OneHotEncoder

!pip install mlxtend==0.23.1

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Association Rule for Store Dataset

In this case study, we will explore how association rule can be used to analyze the items that are usualy purcased together.

you can refer to this article to find out about apriori and association rule:
https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/apriori/
https://rasbt.github.io/mlxtend/user_guide/frequent_patterns/association_rules/

## Load Data

We will use the dataset of the transaction in a certain store. You can get the dataset here:
https://gist.githubusercontent.com/Harsh-Git-Hub/2979ec48043928ad9033d8469928e751/raw/72de943e040b8bd0d087624b154d41b2ba9d9b60/retail_dataset.csv

In [ ]:
# load the data set and show the first five transaction
df = pd.read_csv('https://gist.githubusercontent.com/Harsh-Git-Hub/2979ec48043928ad9033d8469928e751/raw/72de943e040b8bd0d087624b154d41b2ba9d9b60/retail_dataset.csv')
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0       1     2       3       4       5       6
0   Bread    Wine  Eggs    Meat  Cheese  Pencil  Diaper
1   Bread  Cheese  Meat  Diaper    Wine    Milk  Pencil
2  Cheese    Meat  Eggs    Milk    Wine     NaN     NaN
3  Cheese    Meat  Eggs    Milk    Wine     NaN     NaN
4    Meat  Pencil  Wine     NaN     NaN     NaN     NaN

# Get the set of product that has been purchased


In [ ]:
purchased_products = set(item for row in df.values for item in row)
print(purchased_products)

{nan, 'Milk', 'Cheese', 'Wine', 'Diaper', 'Bagel', 'Meat', 'Pencil', 'Bread', 'Eggs'}


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


## Preprocess Data

In this step, we will transform our dataset so that we will have a one hot encoding based on the purchased products.

In [ ]:
# create an itemset based on the products
itemset = set(purchased_products)

# encoding the feature
itemset_encoded = []
for row in df.values:
    row_encoded = {item: 1 if item in row else 0 for item in itemset}
    itemset_encoded.append(row_encoded)

itemset_encoded[0]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


{nan: 0,
 'Milk': 0,
 'Cheese': 1,
 'Wine': 1,
 'Diaper': 1,
 'Bagel': 0,
 'Meat': 1,
 'Pencil': 1,
 'Bread': 1,
 'Eggs': 1}

In [ ]:
# create new dataframe from the encoded features
df_new = pd.DataFrame(itemset_encoded)

# show the new dataframe
df_new.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


NaN  Milk  Cheese  Wine  Diaper  Bagel  Meat  Pencil  Bread  Eggs
0    0     0       1     1       1      0     1       1      1     1
1    0     1       1     1       1      0     1       1      1     0
2    0     1       1     1       0      0     1       0      0     1
3    0     1       1     1       0      0     1       0      0     1
4    0     0       0     1       0      0     1       1      0     0

In [ ]:
# Since, the encoded dataframe consist of the empty column. We will drop the NaN column or u can use the index.
df_new.drop(columns=[float('nan')], inplace=True)
df_new.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Milk  Cheese  Wine  Diaper  Bagel  Meat  Pencil  Bread  Eggs
0     0       1     1       1      0     1       1      1     1
1     1       1     1       1      0     1       1      1     0
2     1       1     1       0      0     1       0      0     1
3     1       1     1       0      0     1       0      0     1
4     0       0     1       0      0     1       1      0     0

Since, the encoded dataframe consist of the empty column. We will drop the NaN column or select all columns other than the first column.

## Apriori Algorithm

We will use appriori algorithm to determine the frequently purchased products.
For this case study, we will min_support=0.2

In [ ]:
#Set threshold value untuk digunakan dalam penghitungan support
from mlxtend.frequent_patterns import apriori, association_rules

min_support = 0.2
frequent_itemsets = apriori(df_new, min_support=min_support, use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support              itemsets
0   0.501587                (Milk)
1   0.501587              (Cheese)
2   0.438095                (Wine)
3   0.406349              (Diaper)
4   0.425397               (Bagel)
5   0.476190                (Meat)
6   0.361905              (Pencil)
7   0.504762               (Bread)
8   0.438095                (Eggs)
9   0.304762        (Milk, Cheese)
10  0.219048          (Milk, Wine)
11  0.225397         (Bagel, Milk)
12  0.244444          (Milk, Meat)
13  0.279365         (Bread, Milk)
14  0.244444          (Milk, Eggs)
15  0.269841        (Wine, Cheese)
16  0.200000      (Diaper, Cheese)
17  0.323810        (Meat, Cheese)
18  0.200000      (Pencil, Cheese)
19  0.238095       (Bread, Cheese)
20  0.298413        (Cheese, Eggs)
21  0.234921        (Diaper, Wine)
22  0.250794          (Meat, Wine)
23  0.200000        (Pencil, Wine)
24  0.244444         (Bread, Wine)
25  0.241270          (Wine, Eggs)
26  0.231746       (Diaper, Bread)
27  0.279365        (Bagel, Bread)
28  0.206349         (Bread, Meat)
29  0.266667          (Meat, Eggs)
30  0.200000       (Pencil, Bread)
31  0.203175  (Milk, Meat, Cheese)
32  0.215873  (Meat, Cheese, Eggs)

Then we will generate association rule of the frequent itemset based on confidence level with the threshold=0.6

In [ ]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6)
rules.drop(columns=['zhangs_metric'], inplace=True, errors='ignore')
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents consequents  antecedent support  consequent support  \
0           (Milk)    (Cheese)            0.501587            0.501587   
1         (Cheese)      (Milk)            0.501587            0.501587   
2           (Wine)    (Cheese)            0.438095            0.501587   
3           (Meat)    (Cheese)            0.476190            0.501587   
4         (Cheese)      (Meat)            0.501587            0.476190   
5           (Eggs)    (Cheese)            0.438095            0.501587   
6          (Bagel)     (Bread)            0.425397            0.504762   
7           (Eggs)      (Meat)            0.438095            0.476190   
8     (Milk, Meat)    (Cheese)            0.244444            0.501587   
9   (Milk, Cheese)      (Meat)            0.304762            0.476190   
10  (Meat, Cheese)      (Milk)            0.323810            0.501587   
11  (Meat, Cheese)      (Eggs)            0.323810            0.438095   
12    (Meat, Eggs)    (Cheese)            0.266667            0.501587   
13  (Eggs, Cheese)      (Meat)            0.298413            0.476190   

     support  confidence      lift  leverage  conviction  
0   0.304762    0.607595  1.211344  0.053172    1.270148  
1   0.304762    0.607595  1.211344  0.053172    1.270148  
2   0.269841    0.615942  1.227986  0.050098    1.297754  
3   0.323810    0.680000  1.355696  0.084958    1.557540  
4   0.323810    0.645570  1.355696  0.084958    1.477891  
5   0.298413    0.681159  1.358008  0.078670    1.563203  
6   0.279365    0.656716  1.301042  0.064641    1.442650  
7   0.266667    0.608696  1.278261  0.058050    1.338624  
8   0.203175    0.831169  1.657077  0.080564    2.952137  
9   0.203175    0.666667  1.400000  0.058050    1.571429  
10  0.203175    0.627451  1.250931  0.040756    1.337845  
11  0.215873    0.666667  1.521739  0.074014    1.685714  
12  0.215873    0.809524  1.613924  0.082116    2.616667  
13  0.215873    0.723404  1.519149  0.073772    1.893773

Provide explanation about __antecedent support__, __consequent support__, __support__, __confidence__, __lift__, __leverage__, __conviction__, __conviction__ and the interpretation from the case above (please use text section)

1. Antecedent Support: This tells you how often the item (or combination of items) on the left side of the rule (antecedent) appears in the dataset.
Example (Milk → Cheese): If the antecedent is Milk, a support of 0.501587 means that Milk appears in about 50.16% of the transactions.
2. Consequent Support: This shows how often the item (or combination of items) on the right side of the rule (consequent) appears in the dataset.
Example (Milk → Cheese): For Cheese, the support of 0.501587 means that Cheese is purchased in 50.16% of all transactions.
3. Support (Rule Support): This measures how often the combination of antecedent and consequent appear together in the dataset.
Example (Milk → Cheese): The support of 0.304762 means that Milk and Cheese together are bought in 30.48% of the transactions.
4. Confidence: Confidence is a measure of the reliability or strength of an association rule. It quantifies the likelihood that if itemset X is purchased, itemset Y will also be purchased. A high confidence value indicates that itemset Y is often purchased when itemset X is purchased, suggesting a strong association or dependency between the two. On the other hand, a low confidence value suggests a weaker association.
Example (Milk → Cheese): A confidence of 0.607595 means that when Milk is bought, there’s about 60.76% chance that Cheese will be bought too.
5. Lift: Lift is a metric that measures the strength of association between two itemsets (A and B) while taking into account the expected or random occurrence of B given A. It is used to determine if the association between A and B is significant beyond what would be expected by chance. Example (Milk → Cheese): A lift of 1.211344 means that Milk and Cheese are 21.13% more likely to be bought together than expected by chance. Lift greater than 1 indicates that the presence of A has a positive effect on the likelihood of B being purchased, suggesting a stronger association between A and B than would be expected by chance. Lift equal to 1 means that there is no association between A and B beyond what would be expected by chance. Lift less than 1 suggests that the presence of A has a negative effect on the likelihood of B being purchased, indicating a weaker association between A and B.
6. Leverage: Leverage measures the difference between the observed frequency of co-occurrence of two items (X and Y) in transactions and the frequency that would be expected if X and Y were independent of each other. In other words, it quantifies how much the occurrence of X and Y together deviates from what would be expected by chance. Example (Milk → Cheese): A leverage of 0.053172 means that Milk and Cheese appear together 5.32% more often than we would expect by chance. A positive leverage value indicates that the occurrence of X and Y together is more frequent than what would be expected by chance, suggesting a positive association. A leverage value of 0 suggests that the co-occurrence of X and Y is as expected under independence, meaning there is no association. A negative leverage value indicates that the occurrence of X and Y together is less frequent than expected, suggesting a negative association or avoidance.
7. Conviction: Conviction measures the ratio of the expected frequency that X occurs without Y to the observed frequency that X occurs without Y. In essence, it quantifies how much the presence of X implies the absence of Y. X high conviction value indicates that the rule X => Y is highly dependent, and A rarely occurs without B. Example (Milk → Cheese): A conviction of 1.270148 means that the rule Milk → Cheese is 1.27 times more likely to be true than if they were independent. A conviction value close to 1 suggests that there is no association between X and Y, as the presence of X does not affect the presence or absence of Y. A high conviction value (> 1) indicates that the presence of X is strongly dependent on the presence of Y, implying a strong association. A low conviction value (< 1) suggests that the presence of X is not strongly dependent on the presence of Y, implying a weaker association.

These metrics help assess the strength and significance of relationships between items, such as Milk → Cheese, in transactional data. Higher values in confidence, lift, and conviction indicate stronger associations between items.

### Sources:


*   https://chatgpt.com/share/674836dd-d980-800c-a059-e1babdd1f1f8
*   https://medium.com/@iuyasik/market-basket-analysis-apriori-algorithm-using-zhangs-metric-708406fc5dfc



